### PR TITLE
planner: fix confused and unnecessary double-projection in plans.

### LINKED ISSUE #20734: confused and unnecessary double-projection in plans
## Development Task
```
mysql> create table t (a int, b int, index(a));
Query OK, 0 rows affected (0.03 sec)

mysql> explain select b from t order by a limit 5;
+------------------------------------+---------+-----------+---------------------+-----------------------------------+
| id                                 | estRows | task      | access object       | operator info                     |
+------------------------------------+---------+-----------+---------------------+-----------------------------------+
| Projection_7                       | 5.00    | root      |                     | test.t.b                          |
| └─Projection_23                    | 5.00    | root      |                     | test.t.a, test.t.b                |
|   └─IndexLookUp_22                 | 5.00    | root      |                     | limit embedded(offset:0, count:5) |
|     ├─Limit_21(Build)              | 5.00    | cop[tikv] |                     | offset:0, count:5                 |
|     │ └─IndexFullScan_19           | 5.00    | cop[tikv] | table:t, index:a(a) | keep order:true, stats:pseudo     |
|     └─TableRowIDScan_20(Probe)     | 5.00    | cop[tikv] | table:t             | keep order:false, stats:pseudo    |
+------------------------------------+---------+-----------+---------------------+-----------------------------------+
6 rows in set (0.00 sec)
```
As shown above, the double-projection is confused and unnecessary, so it's better to merge them into one.

### PR DESCRIPTION
**Pull request by tangwz**
### What problem does this PR solve?

Issue Number: close #20734 

Problem Summary:
Confused and unnecessary double-projection in plans.

### What is changed and how it works?

What's Changed:
Merge projection in phase `eliminatePhysicalProjection`.

How it Works:

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

- None
### Release note 

- No release note

### CODE REVIEW

File: planner/core/rule_eliminate_projection.go [hunk 1]
```diff
@@ -109,6 +118,22 @@ func eliminatePhysicalProjection(p PhysicalPlan) PhysicalPlan {
 type projectionEliminator struct {
 }
 
+// ReplaceColumnOfExprInPhysicalProjection replaces column of expression by another PhysicalProjection.
+func ReplaceColumnOfExprInPhysicalProjection(expr expression.Expression, proj *PhysicalProjection, schema *expression.Schema) expression.Expression {
+	switch v := expr.(type) {
+	case *expression.Column:
```
Comment (qw4990) [inline planner/core/rule_eliminate_projection.go:124]: We should also consider `CorrelatedColumn` here as well.

File: planner/core/rule_eliminate_projection.go [hunk 2]
```diff
@@ -84,8 +84,17 @@ func doPhysicalProjectionElimination(p PhysicalPlan) PhysicalPlan {
 
 	proj, isProj := p.(*PhysicalProjection)
 	if !isProj || !canProjectionBeEliminatedStrict(proj) {
+		if isProj {
```
Comment (qw4990) [inline planner/core/rule_eliminate_projection.go:87]: `if !isProj { if isProj }` seems weird. How about:
```
if isProj {...}
if !isProj ... {...}
```

**Comment by tangwz** [comment]
@qw4990 PTAL.

File: planner/core/rule_eliminate_projection.go [hunk 3]
```diff
@@ -83,9 +83,19 @@ func doPhysicalProjectionElimination(p PhysicalPlan) PhysicalPlan {
...
+				proj.Exprs[i] = expression.FoldConstant(ReplaceColumnOfExprInPhysicalProjection(proj.Exprs[i], child, child.Schema()))
+			}
+			p.Children()[0] = child.Children()[0]
+		}
+		return p
+	}
+	if !isProj {
 		return p
 	}
+
```
Comment (qw4990) [inline planner/core/rule_eliminate_projection.go:98]: How about:
```
	if isProj && canProjectionBeEliminatedStrict(proj) { // eliminate the projection
		return p.Children()[0]
	}
    if isProj {
		if child, ok := p.Children()[0].(*PhysicalProjection); ok && !ExprsHasSideEffects(child.Exprs) { // try to eliminate its child
			for i := range proj.Exprs {
				proj.Exprs[i] = expression.FoldConstant(ReplaceColumnOfExprInPhysicalProjection(proj.Exprs[i], child, child.Schema()))
			}
			p.Children()[0] = child.Children()[0]
            return p
		}
    }
    return p // do nothing
```

File: planner/core/rule_eliminate_projection.go [hunk 4]
```diff
@@ -109,6 +119,24 @@ func eliminatePhysicalProjection(p PhysicalPlan) PhysicalPlan {
 type projectionEliminator struct {
 }
 
+// ReplaceColumnOfExprInPhysicalProjection replaces column of expression by another PhysicalProjection.
+func ReplaceColumnOfExprInPhysicalProjection(expr expression.Expression, proj *PhysicalProjection, schema *expression.Schema) expression.Expression {
```
Comment (qw4990) [inline planner/core/rule_eliminate_projection.go:123]: Should we invoke plan.ResolveIndices() for safety after replacing columns? @winoros

**Review by qw4990 — COMMENTED** [review]
LGTM

File: planner/core/rule_eliminate_projection.go (hunk 4 above)
Comment (winoros) [reply planner/core/rule_eliminate_projection.go:122]: It's ok that the `ResolveIndices` happens later than this.

Comment (winoros) [reply planner/core/rule_eliminate_projection.go:122]: I think we can directly move the `ResolveIndices` after the current post optimize phase.

File: planner/core/rule_eliminate_projection.go [hunk 5]
```diff
@@ -109,6 +118,24 @@ func eliminatePhysicalProjection(p PhysicalPlan) PhysicalPlan {
...
+	switch v := expr.(type) {
+	case *expression.Column:
+		idx := schema.ColumnIndex(v)
+		if idx != -1 && idx < len(proj.Exprs) {
+			return proj.Exprs[idx]
+		}
+	case *expression.CorrelatedColumn:
+		ReplaceColumnOfExprInPhysicalProjection(&v.Column, proj, schema)
+	case *expression.ScalarFunction:
+		for i := range v.GetArgs() {
```
Comment (winoros) [inline planner/core/rule_eliminate_projection.go:132]: Though it's the last phase of optimizing so directly replace the args of one expression should be considered as a safe operation. But we still suggest that create a new expression instead.

**Review by winoros — COMMENTED** [review]
others LGTM, and i just wonder that it's possible to do this in logical optimize phase?

**Comment by tangwz** [comment]
> others LGTM, and i just wonder that it's possible to do this in logical optimize phase?

We can't eliminate it in logical optimize phase because double-proj not appear in the logical optimize phase.

File: planner/core/rule_eliminate_projection.go (hunk 5 above)
Comment (tangwz) [reply planner/core/rule_eliminate_projection.go]: > Though it's the last phase of optimizing so directly replace the args of one expression should be considered as a safe operation. But we still suggest that create a new expression instead.

@winoros PTAL.

**Comment by Reminiscent** [comment]
> > others LGTM, and i just wonder that it's possible to do this in logical optimize phase?
> 
> We can't eliminate it in logical optimize phase because double-proj not appear in the logical optimize phase.

It's possible to not generate anthor proj rather than to eliminate it?

**Comment by ichn-hu** [comment]
any updates on this PR?

**Comment by tangwz** [comment]
@ichn-hu @winoros PTAL.

File: planner/core/rule_eliminate_projection.go [hunk 6]
```diff
@@ -133,12 +155,43 @@ func eliminatePhysicalProjection(p PhysicalPlan) PhysicalPlan {
...
 	}
-	return newRoot
+	return newRoot, nil
 }
 
 type projectionEliminator struct {
 }
 
+// ReplaceColumnOfExprInPhysicalProjection replaces column of expression by another PhysicalProjection.
+func ReplaceColumnOfExprInPhysicalProjection(expr expression.Expression, proj *PhysicalProjection, schema *expression.Schema) (expression.Expression, error) {
```
Comment (tisonkun) [inline planner/core/rule_eliminate_projection.go:165]: Can `ReplaceColumnOfExprInPhysicalProjection` be private?

Comment (tangwz) [reply planner/core/rule_eliminate_projection.go:165]: I agree with you.

Comment (tisonkun) [reply planner/core/rule_eliminate_projection.go]: Thanks for your updates! I am looking for reviewer helping on this PR.

**Comment by tangwz** [comment]
@tisonkun Let me fix this bug later.

**Comment by tisonkun** [comment]
@tangwz isn't it a valid fix now?

**Comment by tangwz** [comment]
@tisonkun No.

**Comment by tisonkun** [comment]
@tangwz thanks for your information. Then I tend to mark this PR as draft.

**Comment by tisonkun** [comment]
Closed as inactive for a while. Feel free to create a PR based on the master.

**Pull request closed without merging**